type-metadata/src/lib.rs

```diff
@@ -1,7 +1,346 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod registry;
+pub use registry::Registry;
+
+mod tests;
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub enum IdentKind {
+	Custom(CustomIdent),
+
+	// primitives, including build-in types and common precludes
+	Bool,
+	Str,
+	U8,
+	U16,
+	U32,
+	U64,
+	U128,
+	I8,
+	I16,
+	I32,
+	I64,
+	I128,
+	Array(ArrayIdent),
+	Slice(SliceIdent),
+	Tuple(TupleIdent),
+	Option(OptionIdent),
+	Result(ResultIdent),
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct Namespace(Vec<&'static str>);
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct CustomIdent {
+	pub name: &'static str,
+	pub namespace: Namespace,
+	pub type_params: Vec<IdentKind>,
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct ArrayIdent {
+	pub len: u16,
+	pub type_param: Box<IdentKind>,
+}
+impl ArrayIdent {
+	fn new(len: u16, type_param: IdentKind) -> Self {
+		ArrayIdent {
+			len,
+			type_param: Box::new(type_param),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct TupleIdent {
+	pub type_params: Option<Vec<IdentKind>>,
+}
+impl TupleIdent {
+	fn new(type_params: Option<Vec<IdentKind>>) -> Self {
+		TupleIdent { type_params }
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct SliceIdent {
+	pub type_param: Box<IdentKind>,
+}
+impl SliceIdent {
+	fn new(type_param: IdentKind) -> Self {
+		SliceIdent {
+			type_param: Box::new(type_param),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct OptionIdent {
+	pub type_param: Box<IdentKind>,
+}
+impl OptionIdent {
+	fn new(type_param: IdentKind) -> Self {
+		OptionIdent {
+			type_param: Box::new(type_param),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct ResultIdent {
+	pub type_params: (Box<IdentKind>, Box<IdentKind>),
+}
+impl ResultIdent {
+	fn new(type_params: (IdentKind, IdentKind)) -> Self {
+		ResultIdent {
+			type_params: (Box::new(type_params.0), Box::new(type_params.1)),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum TypeDef {
+	Primitive,
+	Struct(StructDef),
+	TupleStruct(TupleStructDef),
+	Enum(EnumDef),
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct Field {
+	pub name: &'static str,
+	pub ident: IdentKind,
+}
+#[derive(PartialEq, Eq, Debug)]
+pub struct StructDef(Vec<Field>);
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct TupleStructDef(Vec<IdentKind>);
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct DataVariant<T> {
+	pub name: &'static str,
+	pub index: u16,
+	pub struct_def: T,
```

Comment:
Why is it generic?

Reply:
To avoid duplicated structs, as data variant could have a `StructDef` or `TupleStructDef`:
https://github.com/cennznet/type-metadata/blob/bcec261822515577414afb595d8545bac545156f/type-metadata/src/lib.rs#L134-L138

Reply:
Then maybe some type aliases would make sense here.